[WIP] Getting rid of cross-realm bots

mateuszmandera opened the pull request:
This depends on #17452 and #17516, so it has a bunch of initial commit from those PRs, will rebase appropriately when those get reviewed/fixed and merged. New commits begin with `test_webhooks_common: Compare users by id not email.`

What this does right now: Finishes the first logical part of the migration - adds the bots to every realm, makes `get_system_bots` take a realm (realm_id strictly speaking, to avoid fetching `.realm` attribute of objects just to call this function) argument and adjusts production code and all tests accordingly to make things work and the CI pass. I don't think this should be considered for merging without the follow-up parts. The main commit is `bots: Put system bots in each realm. `, the remaining are just follow-up adjustments to the changes this commit introduces.

What's left to be done:
1. Actually killing cross-realm messages - clean up all the code for handling those (and make sure they can't be sent anymore), rename/remove the various `cross_realm` named concepts.
2. Migrating existing cross-realm messages to use the in-realm bot instead of the `zulipinternal` bot as recipient/sender.

mateuszmandera committed:
bots: Put system bots in each realm.

- migration to add the bots to every existing realm
- code changes to make creation of these bots part of the real creation
  process. This is done via the setup_realm_internal_bots function that
  already existed, but did nothing, because REALM_INTERNAL_BOTS was
  empty. We make these system bots REALM_INTERNAL_BOTS. Nagios bots are
  skipped, as they're not cross-realm bots, they just send messages
  within their zulipinternal realm.
- get_system_bots and friends are adjusted to take a realm argument to
  know from which realm to grab the bot. This makes the functions
  get_user_by_id_in_realm_including_cross_realm and
  get_user_including_cross_realm a bit nonsensical, but we just want to
  make things work for now. Redundant code like this will be cleaned up
  later.

This commit breaks a lot of things, it just lays the foundation for
follow-up changes. All calls to get_system_bot and friends need to now
be adjusted, and various other tweaks will be needed. These will be
addressed in the next commits.

mateuszmandera committed:
bots: Adjust hard-coded id in python_examples.

mateuszmandera committed:
bots: Remove cross-realm bot logic from retention.

We're migrating away from having cross-realm bots messages. The logic
can be entirely removed from retention.

mateuszmandera committed:
bots: Eliminate cross-realm bot messages from export/import system.

mateuszmandera committed:
bots: Add is_system_bot attribute to User zulip.yaml definition.

Now that system bots are in every realm, the is_system_bot attribute
needs to be expected to be found on User objects returnied by the API.

mateuszmandera committed:
bots: Adjust user count in test_stripe.py for the system bot migration.

mateuszmandera committed:
bots: Remove test for get_user_by_id_..._cross_realm in test_users.

The function has become trivial - and is planned for removal in upcoming
commits.

mateuszmandera committed:
bots: Fix remaining tests after system bots migration.

showell commented on zerver/migrations/0311_put_system_bots_in_all_realms.py:
```diff
@@ -0,0 +1,125 @@
+        realm: Any,
+        email: str,
+        full_name: str,
+        password: Optional[str] = None,
```
I wonder if we should remove all the optional parameters here that we don't invoke in the actual migration.

mateuszmandera replied:
True, cleaned it up.

showell commented on zerver/openapi/python_examples.py:
```diff
@@ -84,7 +84,7 @@ def add_subscriptions(client: Client) -> None:
     # To subscribe other users to a stream, you may pass
     # the `principals` argument, like so:
-    user_id = 26
+    user_id = 35
```
Ugh, what does this hard coded id represent?

mateuszmandera replied:
In `create_user` in this file, which is run early on, `newbie@zulip.com` user is created. This example code here wants to subscribe this user to a stream, so it hard-codes the id.

showell commented on zerver/lib/retention.py:
```diff
@@ -251,42 +244,12 @@ def move_expired_personal_and_huddle_messages_to_archive(
         chunk_size=chunk_size,
     )
 
-    # Archive cross-realm personal messages to users in the realm.  We
```
I can't wait to kill off this code! 😄

showell commented:
A good thing to note from Tim when we do final review: "I guess one other thing we'll need to figure out here is how we want administration of these bots to be once they're part of organizations -- have we decided whether administrators will be able to deactivate them (etc.)?"

mateuszmandera commented:
Yeah, that's a good question. I think we should, at least for now, make it impossible to deactivate them (maybe we could allow changing name if that's something useful)- just like right now orgs don't have control over these bots. To allow their deactivation would mean having to take that possibility into account in every place where we try to send a message from a system bot. Conceptually,  the right way  to think of these bots is probably that they're just a built-in, native piece of the server - rather than regular bot account under org administration's control

mateuszmandera committed:
bots: Add migration to turn cross-realm Messages into in-realm ones.

mateuszmandera commented:
Added a db migration that takes all cross-realm messages and "moves them" to belong (whether in the sender or recipient sense) to the corresponding in-realm bot. Tested manually by creating various cross-realm messages, running the migration and then messing around in the Django shell to verify validity of all data (Message senders/recipients, USerMessages, Subscriptions, Huddle hashes). Also by opening the webapp post-migration and verifying things look right.

timabbott commented:
If there are any more prep commits from this that we should look at for 4.0, rebase this and let me know, but I think probably there aren't?

mateuszmandera committed:
curl_examples: Fix hard-coded user ids.

mateuszmandera committed:
puppeteer: Swap ordering of expected recipients in compose.ts.

mateuszmandera committed:
bots: Rename CROSS_REALM_BOT_EMAILS to SYSTEM_BOTS_EMAILS.

mateuszmandera committed:
bots: Rename is_cross_realm_bot_email to is_system_bot.

mateuszmandera committed:
bots: Completely disallow cross-realm messages.

mateuszmandera committed:
javascript_examples: Fix hard-coded user ids.

mateuszmandera committed:
upload_message_file: Remove target_realm argument.

This was only used for uploads done by the email gateway bot. Now that
each realm has its own system bots, and cross-realm messaging is gone,
this code is redundant.

mateuszmandera commented:
@timabbott This should be reviewable now. Here are things that still need to be done:
1. Verify s3 uploads by email gateway bot work (https://chat.zulip.org/#narrow/stream/3-backend/topic/cross.20realm.20bots/near/1125101). Based on my current reading of the code it should work fine without needing upload migrations, but needs empirical verification. This is a blocker for merging.
2. There are still a *lot* of references to cross-realm bots throughout the codebase. This PR is already massive and hard to review, so perhaps those cleanups should be done as follow-ups? Basically, all the references will need to be changed to "system bots" instead - this is easier for comments, but harder for something like `cross_realm_dict` in `people.js` and all the related code which will need more thoughtful renaming of things to make sense. There'll surely also still be chunks of dead code through the codebase that was meant for handling the special cases of cross-realm bots and now will need to be removed.

For help with reviewing - main commits are:
1. `bots: Put system bots in each realm.`
... Then we have a bunch of minor steps to migrate all sorts of pieces of code ...
2. `bots: Add migration to turn cross-realm Messages into in-realm ones.`
3. `bots: Completely disallow cross-realm messages.` is not that large but is the 3rd "core" commit here.

mateuszmandera commented on docs/production/multiple-organizations.md:
```diff
@@ -118,12 +118,7 @@ avoid confusion as to why there's an extra realm when inspecting the
-like "Notification Bot", "Welcome Bot", etc. exist.  In the future,
-we're considering moving these bots to exist in every realm, so that
-we wouldn't need the system realm anymore.
+`zulipinternal` realm.  TODO: What should we write now that cross-realm bots are gone?
```
What do we want to write here? The realm still has some use - some notifications get posted to its streams, e.g this:
```
    # We also send a notification to the Zulip administrative realm
    admin_realm = get_realm(settings.SYSTEM_BOT_REALM)
    admin_realm_sender = get_system_bot(sender_email, admin_realm.id)
    try:
        # Check whether the stream exists
        signups_stream = get_signups_stream(admin_realm)
        # We intentionally use the same strings as above to avoid translation burden.
        message = _("{user} just signed up for Zulip. (total: {user_count})").format(
            user=f"{user_profile.full_name} <`{user_profile.email}`>", user_count=user_count
        )
        internal_send_stream_message(
            admin_realm_sender, signups_stream, user_profile.realm.display_subdomain, message
        )

    except Stream.DoesNotExist:
        # If the signups stream hasn't been created in the admin
        # realm, don't auto-create it to send to it; just do nothing.
        pass
```

timabbott replied:
I think probably the right model is to have the `SYSTEM_BOT_REALM` setting control which realm is used for that, but have the default be that it's unset and we don't attempt to send those notifications.

timabbott commented:
I merged the couple commits ending with 119f1da04afba9ac3b7f3cc1287597e981955b6b, after adding an API feature level bump in that second commit. 

My recommended next step is to make a separate PR on top of master for the commits that pass a `realm_id` to `get_system_bot`; we can merge those all as preparatory commits with `get_system_bot` ignoring the `realm_id` parameter, to get that refactoring in without changing any production behavior.  We can then lock that down by making the realm a required parameter to `get_system_bot`, and have most of the "individual feature" audit work fully integrated in master before we try to do the "big cutover" commit.

mateuszmandera committed:
models: Rename get_user_including_cross_realm to ..._system_bots.

mateuszmandera committed:
models: Change get_user_..._cross_realm  to exclude cross-realm.

With cross-realm bots eliminated, the function should work differently.
Now it should just return the user with specified id if the user is in
the correct realm.

mateuszmandera commented:
@timabbott This is now rebased and should be reviewable again. This doesn't have the removal system bot realm  or the other changes to go with it, but I think that may be best as a direct follow-up to merging the current part given how large it already is? This does leave the system bot realm in a pretty useless state, but shouldn't break anything.

mateuszmandera committed:
bots: Disallow modification of system bots by realm admins.
